### PR TITLE
Workload and Namespace recommendation

### PR DESCRIPTION
**Pull request by gizzi-rh**

### CODE REVIEW

**Comment by gizzi-rh** [comment]
Hi team, I am available if you need any info on my side

File: batch_worker.py [hunk 1]
```diff
@@ -1,257 +1,41 @@
...
-                self.endpointLogger.error(
-                    "Only exclude/include actions are allowed for namespace filter."
-                    f" Using filter {namespace_filter_str}"
-                )
+        self.start_date = start_date 
+        self.end_date = end_date 
+        self.tolerance = tolerance 
+        self.batch_size = batch_size
+
+    def start_work(self):
```
Comment (dvandra) [inline batch_worker.py:29]: rename to `start_process_metric_data`

File: recommender.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    def get_cluster_id_batch(self, end_date, batch_size):
+        cluster_ids = self.get_metric_data(CLUSTER_IDS, end_date.strftime("%s"))[ID_LABEL].unique()
+        print("Reading clusters of batch size : "+ str(batch_size))
+        result = [
+            "|".join(cluster_ids[pos:pos + batch_size])
+            for pos in range(0, len(cluster_ids), batch_size)
+        ]
+        return result
+
+    #def get_grafana_url_for_namespace(self, df):
```
Comment (dvandra) [inline recommender.py:198]: No need to remove this function here:  it will get grafana_url_for_namespaces.

File: recommender.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            ["dummy_counter_x", "dummy_counter_y"]
+        ].fillna(0).apply(min, axis=1)
+        df_merged.drop(columns=["dummy_counter_x", "dummy_counter_y"], inplace=True)
+        
+        self.endpointLogger.info("Output for endpoint %s : %s",self.name,self.endpoint)
+        # Sort By Delta
+        df_merged.drop(columns=['ephemeral_storage_value'],errors='ignore',inplace=True)
+        df_merged.sort_values(by=['cpu_delta','memory_delta'],ascending=False,na_position="last",inplace=True)
+        #df_merged['grafana_url'] = self.endpoint['url'].replace("rbac-query-proxy","grafana")
+        #df_merged['grafana_url'] = df_merged.apply(self.get_grafana_url_for_namespace, axis=1)
```
Comment (dvandra) [inline recommender.py:245]: No need to remove this here: it will get grafana_url_for_namespaces.

File: start.py [hunk 4]
```diff
@@ -37,13 +38,19 @@ def process_input():
...
             print("Processing acm hub : ", item['hub_name'] , "on past",days_delta ,"days of metrics")
+            
+            worker=batch_worker(conn,item['hub_name'],recommender_type, item,grafana_dashboard_uid, start_date, end_date, tolerance, batch_size)
+            worker.start_work()
+
+            """
             if grafana_dashboard_uid:
-                worker=batch_worker(conn,item['hub_name'],item,grafana_dashboard_uid)
-                worker.process_metric_data(start_date, end_date, tolerance,batch_size)
+                worker=batch_worker(conn,item['hub_name'],recommender_type, item,grafana_dashboard_uid, start_date, end_date, tolerance, batch_size)
```
Comment (dvandra) [inline start.py:47]: No need to remove this here: it will get_grafana_url_for_namespaces.

File: batch_worker.py [hunk 5]
```diff
@@ -1,257 +1,41 @@
...
+        if self.recommender_type == WORKLOAD_RECOMMENDER_TYPE:
+            recommender = WorkloadRecommender(self.name, self.p_client, self.endpoint, self.start_date, self.end_date, self.tolerance, self.batch_size)
```
Comment (dvandra) [inline batch_worker.py:31]: add `grafana_dashboard_uid` as a parameter.

File: batch_worker.py [hunk 6]
```diff
@@ -1,257 +1,41 @@
...
-    def get_usage_data_batch(self,start_date, end_date, cluster_ids_filter):
-        cpu_usage_max = []
-        memory_usage_max = []
-        cpu_usage_query = CPU_USAGE_MAX.replace(
-            "NAMESPACE_FILTER", self._namespace_filter
-        )
-        mem_usage_query = MEMORY_USAGE_MAX.replace(
-            "NAMESPACE_FILTER", self._namespace_filter
-        )
+            recommender = NamespaceRecommender(self.name, self.p_client, self.endpoint, self.start_date, self.end_date, self.tolerance, self.batch_size)
```
Comment (dvandra) [inline batch_worker.py:33]: add `grafana_dashboard_uid` as a parameter.

File: namespace_recommender.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+class NamespaceRecommender(Recommender):
+
+    key = ["cluster", "namespace"]
+    requestQuery = RESOURCE_REQUEST_NAMESPACE
+    cpuMaxQuery = CPU_USAGE_MAX_NAMESPACE
+    memoryMaxQuery = MEMORY_USAGE_MAX_NAMESPACE
+
+    def __init__(self, name, prometheus_connection, endpoint, start_date, end_date, tolerance, batch_size) -> None:
+        super().__init__(name, prometheus_connection, endpoint, start_date, end_date, tolerance, batch_size)
```
Comment (dvandra) [inline namespace_recommender.py:12]: add `grafana_dashboard_uid` as a parameter.

File: workload_recommender.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+class WorkloadRecommender(Recommender):
+
+    key = ["cluster", "namespace", "workload"]
+    requestQuery = RESOURCE_REQUEST_WORKLOAD
+    cpuMaxQuery = CPU_USAGE_MAX_WORKLOAD
+    memoryMaxQuery = MEMORY_USAGE_MAX_WORKLOAD
+
+
+    def __init__(self, name, prometheus_connection, endpoint, start_date, end_date, tolerance, batch_size) -> None:
+        super().__init__(name, prometheus_connection, endpoint, start_date, end_date, tolerance, batch_size)
```
Comment (dvandra) [inline workload_recommender.py:16]: add `grafana_dashboard_uid` as a parameter.

File: recommender.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
+from logger import *
+from queries import *
+from pandas import DataFrame,concat,date_range
+from timeit import default_timer as timer
+from datetime import datetime
+
+class Recommender:
+
+    def __init__(self, name, prometheus_connection, endpoint, start_date, end_date, tolerance, batch_size ) -> None:
```
Comment (dvandra) [inline recommender.py:9]: add `grafana_dashboard_uid` as a parameter.

File: recommender.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        self.name=name
+        self.prometheus_connection=prometheus_connection
+        self.endpoint=endpoint
```
Comment (dvandra) [inline recommender.py:12]: add `self.grafana_dashboard_uid` here.

**Comment by dvandra** [comment]
Please rebase with the latest code.

**Comment by dvandra** [comment]
PTAL: https://github.com/gizzi-rh/thanos-metrics-analyzer/pull/1

**Comment by gizzi-rh** [comment]
@dvandra done it!

**Review by dvandra — APPROVED** [review]
LGTM

**Pull request merged**